src/Symfony/Bridge/Twig/Extension/RoutingExtension.php

```diff
@@ -42,6 +45,7 @@ public function getFunctions()
         return array(
             new TwigFunction('url', array($this, 'getUrl'), array('is_safe_callback' => array($this, 'isUrlGenerationSafe'))),
             new TwigFunction('path', array($this, 'getPath'), array('is_safe_callback' => array($this, 'isUrlGenerationSafe'))),
+            new TwigFunction('has_route', array($this, 'hasRoute')),
```

Comment:
I don't think this should be possible in twig. Generating routes is one thing, but giving the developer the ability to check if a route exists, sounds like a poor solution to see if certain features are available. I think it's better to toggle those features using different means, such as a feature bundle or true/false coming from the controller.

Reply:
if it's possible in a controller, it should be in twig also to me, allowing building flexible systems.

Reply:
What's the relation to soap? We can go in very lengthy unrelated comparisons without making any progress on this one for sure...
This PR is about #19273 (link missing in the description btw!)

Reply:
The idea of templating is to contain as little logic as possible (soap call, check if route exists, call something from a repository). If there is any, it should be hidden from the implementation and only ask whether something should be done or not. In this scenario that would mean that you want to do X or Y based on whether a route exists, this logic should be in your controller and tell your template what to do.

 I have no objections to making this available in the router and if developers wish to implement this twig function in their project, that's their choice. I don't think this should be in the Symfony core as the framework should promote correct usage of templating. The AbstractController could even support a shortcut to make it easier to use.

```php
FooController extends AbstractController
{
    public function fooAction()
    {
        return $this->render('foo.html.twig', [
            'should_show_link_x' => $this->routeExists('app.route_x'),
        ]);
    }
}
```
```twig
{% if should_show_link_x %}
    <a href="{{ path('app.route_x') }}">Route X</a>
{% endif %}

{# vs #} 
{% if has_route('app.route_x') %}
    <a href="{{ path('app.route_x') }}">Route X</a>
{% endif %}
```

In normal scenarios, you control the routes, you know which routes exist, so to me this looks like an extension point in bundles or some form of modules. I don't think this is the right way of checking this extension point. The presence of a route should not define whether a specific feature is enabled or not. Not only is this very fragile, it also means that during container compilation you don't know whether or not a specific feature will be present, making it an inefficient way of enabling or disabling features.

Reply:
Note that this PR is also about adding `AbstractController::hasRoute()` (same as your "routeExists" proposal) and doesn't add anything to the router (it already exposes all that is needed.)

I'm all with you on making templates embed little logic.
Yes I see little difference between `has_route()` and `is_granted()` here: if the later is fone in core, I don't see why the former shouldn't.

Reply:
`is_granted()` is calling the abstraction layer to grant permissions, while `has_route()` is a direct check to see if a route exists, there is no abstraction at play here. If it was `is_feature_enabled('foo')`, it would be a different story. If someone adds a route "because it doesn't exist and this link is not displayed", there's no question and it can just be added. There won't be anyone asking "why is this feature not present, how do I get this feature?".

Reply:
@voronkovich maybe you'd like to chime in here? WDYT?

Reply:
@nicolas-grekas, I agree with @iltar. I think we don't need a twig function. Having the `routeExists` method in the controller is enough.

Reply:
Fine to me then, thanks.